optimum/habana/transformers/models/gptj/modeling_gptj.py

```diff
@@ -121,10 +221,29 @@ def forward(
                 key = torch.cat([past_key, key], dim=-2)
                 value = torch.cat([past_value, value], dim=-2)
 
-        if use_cache is True:
+        if use_cache is True and token_idx is not None:
+            if reuse_cache:
+                key = self.k_cache(key, 2, token_idx)
+                value = self.v_cache(value, 2, token_idx)
+                present = (self.k_cache.get_shape(), self.v_cache.get_shape())
+            else:
+                if layer_past is None:
+                    past_key = torch.zeros(key.shape, dtype=self.k_proj.weight.dtype, device=key.device)
+                    past_value = torch.zeros(key.shape, dtype=self.k_proj.weight.dtype, device=key.device)
+                    layer_past = (past_key, past_value)
+                key = self.k_cache.update(layer_past[0], key, 2, token_idx, self.inp_seq_len)
+                value = self.v_cache.update(layer_past[1], value, 2, token_idx, self.inp_seq_len)
+                present = layer_past
+
+            if cache_idx is not None and q_len == 1:
+                key = key[:, :, :cache_idx, :]
+                value = value[:, :, :cache_idx, :]
+                if attention_mask is not None:
+                    attention_mask = attention_mask[:, :, :, :cache_idx]
+
             # Note that this cast is quite ugly, but is not implemented before ROPE as the original codebase keeps the key in float32 all along the computation.
             # Reference: https://github.com/kingoflolz/mesh-transformer-jax/blob/f8315e3003033b23f21d78361b288953064e0e76/mesh_transformer/layers.py#L128
```

Comment:
```suggestion
```
Can we remove this? It is not relevant anymore

Reply:
done